File: 1584-min-cost-to-connect-all-points/1584-min-cost-to-connect-all-points.cpp
class Solution {
public:
    struct mycomp
    {
        bool operator()(const array<int,2>&a, const array<int,2>&b)
        {
            return a[1]>b[1];
        }
    };
    
    int minCostConnectPoints(vector<vector<int>>& points) {
       
        int n = points.size();
        vector<vector<array<int,2>>>adj(n);
        
        for(int i=0;i<n;i++)
        {
            for(int j=i+1;j<n;j++)
            {
                int dist = abs(points[i][0] - points[j][0]) + abs(points[j][1] - points[i][1]);
                adj[j].push_back({i,dist});
                adj[i].push_back({j,dist});
            }
        }
        
        priority_queue<array<int,2>, vector<array<int,2>> , mycomp>pq;
        pq.push({0,0});
        
        vector<bool>vis(n,false);
        int count =0, cost=0;
                
        while(pq.size()>0 and count<n)
        {
            auto p = pq.top();
            pq.pop();
            
            if(vis[p[0]])
                continue;
            
            vis[p[0]]=true;
            count++;            
            cost+=p[1];
            
            for(auto &neigh : adj[p[0]])
            {
                if(!vis[neigh[0]])
                    pq.push(neigh);
            }
        }
        
        return cost;
    }
};

File: 1584-min-cost-to-connect-all-points/1584-min-cost-to-connect-all-points.java
class Solution {
    
    private ArrayList<int[]> edges;
    private int n;
    private int[] par, rank;

    
    private int findPar(int p) {
        if (par[p] == p)
            return par[p];
        return par[p] = findPar(par[p]);
    }

    
    private int DSUnion(int[] edge) {
        int dist = edge[0], u = edge[1], v = edge[2];
        int p1 = findPar(u), p2 = findPar(v);
        
        if (p1 == p2)
            return 0;

        if (rank[p1] >=rank[p2]) {
            par[p2] = p1;
            rank[p1]++;
        } else {
            par[p1] = p2;
            rank[p2]++;
        }

        return dist;
    }

    public int minCostConnectPoints(int[][] points) {
        this.n = points.length;
        this.edges = new ArrayList<>();

        
        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n; j++) {
                int x = Math.abs(points[i][0] - points[j][0]);
                int y = Math.abs(points[i][1] - points[j][1]);
                int dist = x + y;
                edges.add(new int[]{dist, i, j});
            }
        }

    
        this.par = new int[n];
        this.rank = new int[n];
        for (int i = 0; i < n; i++) {
            par[i] = i;
            rank[i] = 0;
        }


        Collections.sort(edges, (a, b) -> Integer.compare(a[0], b[0]));

        int ans = 0;


        for (int i = 0; i < edges.size(); i++) {
            ans += DSUnion(edges.get(i));
        }

        return ans;
    }
}